File: app/src/main/java/com/falcon/restaurants/helpers/ImagePicker.java
package com.falcon.restaurants.helpers;
import android.annotation.SuppressLint;
import android.content.Context;
import android.net.Uri;
import com.qingmei2.rximagepicker.core.RxImagePicker;
import com.qingmei2.rximagepicker.entity.Result;
import com.qingmei2.rximagepicker.entity.sources.Camera;
import com.qingmei2.rximagepicker.entity.sources.Gallery;
import javax.inject.Inject;
import io.reactivex.Observable;
import io.reactivex.functions.Consumer;

public class ImagePicker {

    Context context;

    @Inject
    public ImagePicker(Context context) {
        this.context = context;
    }

    public interface Listener{
        void onSuccess(Uri uri);
    }

    @SuppressLint("CheckResult")
    public void pick(Listener listener){

        RxImagePicker
                .create(MyImagePicker.class)
                .openGallery(context)
                .subscribe(new Consumer<Result>() {
                    @Override
                    public void accept(Result result) throws Exception {
                        listener.onSuccess(result.getUri());
                    }
                });
    }

    public interface MyImagePicker {
        @Gallery
        Observable<Result> openGallery(Context context);

        @Camera
        Observable<Result> openCamera(Context context);
    }

}

File: app/src/main/java/com/falcon/restaurants/helpers/ImagePicker.kt
package com.falcon.restaurants.helpers
import android.annotation.SuppressLint
import android.content.Context
import android.net.Uri
import com.qingmei2.rximagepicker.core.RxImagePicker
import com.qingmei2.rximagepicker.entity.Result
import com.qingmei2.rximagepicker.entity.sources.Camera
import com.qingmei2.rximagepicker.entity.sources.Gallery
import javax.inject.Inject
import io.reactivex.Observable
import io.reactivex.functions.Consumer

class ImagePicker @Inject constructor (val context: Context) {

    interface Listener{
        fun onSuccess(uri: Uri)
    }

    @SuppressLint("CheckResult")
    fun pick(listener: Listener){

            RxImagePicker
                .create(MyImagePicker::class.java)
                .openGallery(context)
                .subscribe(object :Consumer<Result> {
                    @Throws(Exception::class)
                    override fun accept(result: Result) {
                        listener.onSuccess(result.uri)
                    }
                })
    }

    interface MyImagePicker {
        @Gallery
        fun openGallery(context: Context): Observable<Result>
        @Camera
        fun openCamera(context: Context): Observable<Result>
    }

}
